// memoizee/lib/configure-map-compiled.js
'use strict';

var customError = require('es5-ext/error/custom'),
    defineLength = require('es5-ext/function/_define-length'),
    d = require('d'),
    ee = require('event-emitter').methods,
    resolveResolve = require('./resolve-resolve'),
    resolveNormalize = require('./resolve-normalize'),
    apply = Function.prototype.apply,
    call = Function.prototype.call,
    create = Object.create,
    hasOwnProperty = Object.prototype.hasOwnProperty,
    defineProperties = Object.defineProperties,
    _on = ee.on,
    emit = ee.emit;

module.exports = function (original, length, options) {
	var cache = create(null),
	    conf,
	    memLength,
	    _get,
	    set,
	    del,
	    _clear,
	    extDel,
	    normalizer,
	    getListeners,
	    setListeners,
	    deleteListeners,
	    memoized,
	    resolve;
	if (length !== false) memLength = length;else if (isNaN(original.length)) memLength = 1;else memLength = original.length;

	if (options.normalizer) {
		normalizer = resolveNormalize(options.normalizer);
		_get = normalizer.get;
		set = normalizer.set;
		del = normalizer['delete'];
		_clear = normalizer.clear;
	}
	if (options.resolvers != null) resolve = resolveResolve(options.resolvers);

	if (_get) {
		memoized = defineLength(function (arg) {
			var id,
			    result,
			    args = arguments;
			if (resolve) args = resolve(args);
			id = _get(args);
			if (id !== null) {
				if (hasOwnProperty.call(cache, id)) {
					if (getListeners) conf.emit('get', id, args, this);
					return cache[id];
				}
			}
			if (args.length === 1) result = call.call(original, this, arg);else result = apply.call(original, this, args);
			if (id === null) {
				id = _get(args);
				if (id !== null) throw customError("Circular invocation", 'CIRCULAR_INVOCATION');
				id = set(args);
			} else if (hasOwnProperty.call(cache, id)) {
				throw customError("Circular invocation", 'CIRCULAR_INVOCATION');
			}
			cache[id] = result;
			if (setListeners) conf.emit('set', id);
			return result;
		}, memLength);
	} else if (length === 0) {
		memoized = function () {
			var result;
			if (hasOwnProperty.call(cache, 'data')) {
				if (getListeners) conf.emit('get', 'data', arguments, this);
				return cache.data;
			}
			if (!arguments.length) result = call.call(original, this);else result = apply.call(original, this, arguments);
			if (hasOwnProperty.call(cache, 'data')) {
				throw customError("Circular invocation", 'CIRCULAR_INVOCATION');
			}
			cache.data = result;
			if (setListeners) conf.emit('set', 'data');
			return result;
		};
	} else {
		memoized = function (arg) {
			var result,
			    args = arguments,
			    id;
			if (resolve) args = resolve(arguments);
			id = String(args[0]);
			if (hasOwnProperty.call(cache, id)) {
				if (getListeners) conf.emit('get', id, args, this);
				return cache[id];
			}
			if (args.length === 1) result = call.call(original, this, args[0]);else result = apply.call(original, this, args);
			if (hasOwnProperty.call(cache, id)) {
				throw customError("Circular invocation", 'CIRCULAR_INVOCATION');
			}
			cache[id] = result;
			if (setListeners) conf.emit('set', id);
			return result;
		};
	}
	conf = {
		original: original,
		memoized: memoized,
		get: function get(args) {
			if (resolve) args = resolve(args);
			if (_get) return _get(args);
			return String(args[0]);
		},
		has: function has(id) {
			return hasOwnProperty.call(cache, id);
		},
		'delete': function _delete(id) {
			var result;
			if (!hasOwnProperty.call(cache, id)) return;
			if (del) del(id);
			result = cache[id];
			delete cache[id];
			if (deleteListeners) conf.emit('delete', id, result);
		},
		clear: function clear() {
			var oldCache = cache;
			if (_clear) _clear();
			cache = create(null);
			conf.emit('clear', oldCache);
		},
		on: function on(type, listener) {
			if (type === 'get') getListeners = true;else if (type === 'set') setListeners = true;else if (type === 'delete') deleteListeners = true;
			return _on.call(this, type, listener);
		},
		emit: emit,
		updateEnv: function updateEnv() {
			original = conf.original;
		}
	};
	if (_get) {
		extDel = defineLength(function (arg) {
			var id,
			    args = arguments;
			if (resolve) args = resolve(args);
			id = _get(args);
			if (id === null) return;
			conf['delete'](id);
		}, memLength);
	} else if (length === 0) {
		extDel = function () {
			return conf['delete']('data');
		};
	} else {
		extDel = function (arg) {
			if (resolve) arg = resolve(arguments)[0];
			return conf['delete'](arg);
		};
	}
	defineProperties(memoized, {
		__memoized__: d(true),
		'delete': d(extDel),
		clear: d(conf.clear)
	});
	return conf;
};
